stdlib/math.ysh

```diff
@@ -1,79 +1,79 @@
+func _select(list, cmp) {
+  ## Internal helper for `max` and `min`.
+  ##
+  ## NOTE: If `list` is empty, then an error is thrown.
 
+  if (len(list) === 0) {
+    error ("Unexpected empty list", status=3)
+  }
 
-# We recommend abs(float1 - float2) < 0.0001
-# instead of exact float equality
+  if (len(list) === 1) {
+    return (list[0])
+  }
 
-# This causes a name collision with `abs`. The rename is a work around for now.
-func abs_(x) {
-  if (x < 0) {
-    return (-x)
-  } else {
-    return (x)
+  var match = list[0]
+  for i in (1:len(list)) {
+    setvar match = cmp(list[i], match)
   }
+  return (match)
 }
 
-
-# How to implement Python-like max in YSH?
-
-# max(5,6)  => 6
-
-# max([5,6,7])  => 7
-
-# max(5,6,7,8)  => 8  - I don't like this about Python
-
-func mymax (...args) {
-  # Could also write this
-  # case (len(args)) {
-  #   (0) { error }
-  #   (1) {
-  #     ...
-  #   }
-  #   (2) {
-  #     ...
-  #   }
-  #   (else) { error }
-  # }
-
-  if (len(args) === 0) {
-    error ('Requires 1 arg')
-
-  } elif (len(args) === 1) {
-    # assert List
-
-    var mylist = args[0]
-
-    var max = mylist[0]
-
-    # Could be this, but it's wasteful
-    #for item in (mylist[1:]) {
-
-    for item in (mylist) {
-      if (item > max) {
-        setvar max = item
+func max(...args) {
+  ## Compute the maximum of 2 or more values.
+  ##
+  ## `max` takes two different signatures:
+  ##  - `max(a, b)` to return the maximum of `a`, `b`
+  ##  - `max(list)` to return the greatest item in the `list`
+  ##
+  ## So, for example:
+  ##
+  ##   max(1, 2)  # => 2
+  ##   max([1, 2, 3])  # => 3
+
+  case (len(args)) {
+    (1) { return (_select(args[0], max)) }
+    (2) {
+      if (args[0] > args[1]) {
+        return (args[0])
+      } else {
+        return (args[1])
       }
     }
+    (else) { error ("max expects 1 or 2 args") }
```

Comment:
It's kinda weird we get the status 1 here, but we have explicit status 3 elsewhere

Maybe we should change the default `error` status to 3 ?

Or `4` ?  

Just because it's different than say `ls` failures which are 1 or 2

I guess the issue is that the builtins in Python will `raise error.InvalidType()`, which gives a 3 status code

So then it becomes harder to get consistent behavior between builtins implemented in Python vs YSH

Reply:
AFAIK error code `1` is used as a catch-all error code, so maybe it's fine left as the default?

But you are right, we should use a status `3` here to behave like the `raise error.InvalidType()` that we threw in the old builtins.